xds/internal/client/client_requests_counter_test.go

```diff
@@ -0,0 +1,161 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package client
+
+import (
+	"sync"
+	"sync/atomic"
+	"testing"
+)
+
+type counterTest struct {
+	name          string
+	maxRequests   uint32
+	numRequests   uint32
+	errorExpected bool
+}
+
+var tests = []counterTest{
+	{
+		name:          "does-not-exceed-max-requests",
+		maxRequests:   1024,
+		numRequests:   1024,
+		errorExpected: false,
+	},
+	{
+		name:          "exceeds-max-requests",
+		maxRequests:   32,
+		numRequests:   64,
+		errorExpected: true,
+	},
+}
+
+func counterTestInit() {
+	src = &servicesRequestsCounter{
+		services: make(map[string]*ServiceRequestsCounter),
+	}
+}
+
+func testCounter(t *testing.T, test counterTest) {
+	SetMaxRequests(test.name, &test.maxRequests)
+	requestsStarted := make(chan struct{})
+	requestsSent := sync.WaitGroup{}
+	requestsSent.Add(int(test.numRequests))
+	requestsDone := sync.WaitGroup{}
+	requestsDone.Add(int(test.numRequests))
+	var lastError atomic.Value
+	var successes, errors uint32
+	for i := 0; i < int(test.numRequests); i++ {
+		go func() {
+			counter := GetServiceRequestsCounter(test.name)
+			defer requestsDone.Done()
+			err := counter.StartRequest()
+			if err == nil {
+				atomic.AddUint32(&successes, 1)
+			} else {
+				atomic.AddUint32(&errors, 1)
+				lastError.Store(err)
+			}
+			requestsSent.Done()
+			if err == nil {
+				<-requestsStarted
+				counter.EndRequest()
+			}
+		}()
+	}
+	requestsSent.Wait()
+	close(requestsStarted)
+	requestsDone.Wait()
+	loadedError := lastError.Load()
+	if test.errorExpected && loadedError == nil {
+		t.Error("no error when error expected")
+	}
+	if !test.errorExpected && loadedError != nil {
+		t.Errorf("error starting request: %v", loadedError.(error))
```

Comment:
I don't think `loadedError.(error)` works. It would probably panic. This test works because that if is never true.

Reply:
I tried adding the following and it did print in the case where it exceeds.

```
if loadedError != nil {
	fmt.Println(loadedError.(error))
}
```